File: jans-config-api/common/src/main/java/io/jans/configapi/model/configuration/ApiAppConfiguration.java
package io.jans.configapi.model.configuration;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import io.jans.as.model.configuration.Configuration;
import io.jans.configapi.util.ApiConstants;

import java.util.List;

import io.swagger.v3.oas.annotations.media.Schema;

@JsonIgnoreProperties(ignoreUnknown = true)
public class ApiAppConfiguration implements Configuration {
    
    @Schema(description = "Config API service name.")
    private String serviceName;

    @Schema(description = "OAuth authentication enable/disable flag. Default value `true`.")
    private boolean configOauthEnabled;

    @Schema(description = "Flag to enable/disable timer to dynamically reflect log configuration changes. Default value `true`Default value `false`.")
    private boolean disableLoggerTimer;

    @Schema(description = "Flag to enable/disable request audit. Default value `false`.")
    private boolean disableAuditLogger;

    @Schema(description = "Flag to enable/disable check if custom attribue is declared in schema. Default value `true`.")
    private boolean customAttributeValidationEnabled;

    @Schema(description = "Flag to enable/disable check if acr customScript is enabled. Default value `true`.")
    private boolean acrValidationEnabled;

    @Schema(description = "List of approved external Auth server to validate token.")
    private List<String> apiApprovedIssuer;

    @Schema(description = "Name of supported API protection mechansim. Supported type is `OAuth2`.")
    private String apiProtectionType;

    @Schema(description = "Config-API client ID.")
    private String apiClientId;

    @Schema(description = "Config-API client password.")
    private String apiClientPassword;

    private boolean endpointInjectionEnabled;

    @Schema(description = "Issuer Identifier of Jans OpenID Connect Provider.")
    private String authIssuerUrl;

    @Schema(description = "Jans OpenID Connect Provider Well-Known Configuration URL.")
    private String authOpenidConfigurationUrl;

    @Schema(description = "Jans URL of the OpenID Connect Provider's OAuth 2.0 Authorization Endpoint.")
    private String authOpenidIntrospectionUrl;

    @Schema(description = "Jans URL of the OpenID Connect Provider's OAuth 2.0 Token Endpoint.")
    private String authOpenidTokenUrl;

    @Schema(description = "Jans URL of the OpenID Connect Provider's OAuth 2.0 Revoke Token Endpoint.")
    private String authOpenidRevokeUrl;

    @Schema(description = "List of oAuth scope that can be validity for an access tokens only by underlying Jans Auth server.")
    private List<String> exclusiveAuthScopes;

    @Schema(description = "CORS configuration filter properties.")
    private List<CorsConfigurationFilter> corsConfigurationFilters;

    @Schema(description = "Specify logging level of Loggers. Default level is `INFO`.")
    private String loggingLevel;

    @Schema(description = "Log4j logging layout. Default value `TEXT`.")
    private String loggingLayout;

    @Schema(description = "The path to the external log4j2 logging configuration.")
    private String externalLoggerConfiguration;

    @Schema(description = "Choose whether to disable JDK loggers.")
    private Boolean disableJdkLogger = true;

    @Schema(description = "Maximum number of results per page in search endpoints.")
    private int maxCount;

    @Schema(description = "List of ACR values that should be excluded from active validation check.")
    private List<String> acrExclusionList;
    
    @Schema(description = "User attribute that should not be returned in response.")
    private List<String> userExclusionAttributes;

    @Schema(description = "List of User mandatory attribute for user creation request.")
    private List<String> userMandatoryAttributes;

    @Schema(description = "Agama configuration details.")
    private AgamaConfiguration agamaConfiguration;

    @Schema(description = "Audit Log configuration details.")
    private AuditLogConf auditLogConf;

    @Schema(description = "Configuration for data-type converstion.")
    private DataFormatConversionConf dataFormatConversionConf;

    @Schema(description = "Details of enabled plugins.")
    private List<PluginConf> plugins;

    @Schema(description = "Asset management configuration details.")
    private AssetMgtConfiguration assetMgtConfiguration;

    public String getServiceName() {
        return serviceName;
    }

    public void setServiceName(String serviceName) {
        this.serviceName = serviceName;
    }

    public boolean isConfigOauthEnabled() {
        return configOauthEnabled;
    }

    public void setConfigOauthEnabled(boolean configOauthEnabled) {
        this.configOauthEnabled = configOauthEnabled;
    }

    public boolean isDisableLoggerTimer() {
        return disableLoggerTimer;
    }

    public void setDisableLoggerTimer(boolean disableLoggerTimer) {
        this.disableLoggerTimer = disableLoggerTimer;
    }

    public boolean isDisableAuditLogger() {
        return disableAuditLogger;
    }

    public void setDisableAuditLogger(boolean disableAuditLogger) {
        this.disableAuditLogger = disableAuditLogger;
    }

    public boolean isCustomAttributeValidationEnabled() {
        return customAttributeValidationEnabled;
    }

    public void setCustomAttributeValidationEnabled(boolean customAttributeValidationEnabled) {
        this.customAttributeValidationEnabled = customAttributeValidationEnabled;
    }

    public boolean isAcrValidationEnabled() {
        return acrValidationEnabled;
    }

    public void setAcrValidationEnabled(boolean acrValidationEnabled) {
        this.acrValidationEnabled = acrValidationEnabled;
    }

    public List<String> getApiApprovedIssuer() {
        return apiApprovedIssuer;
    }

    public void setApiApprovedIssuer(List<String> apiApprovedIssuer) {
        this.apiApprovedIssuer = apiApprovedIssuer;
    }

    public String getApiProtectionType() {
        return apiProtectionType;
    }

    public void setApiProtectionType(String apiProtectionType) {
        this.apiProtectionType = apiProtectionType;
    }

    public String getApiClientId() {
        return apiClientId;
    }

    public void setApiClientId(String apiClientId) {
        this.apiClientId = apiClientId;
    }

    public String getApiClientPassword() {
        return apiClientPassword;
    }

    public void setApiClientPassword(String apiClientPassword) {
        this.apiClientPassword = apiClientPassword;
    }

    public boolean isEndpointInjectionEnabled() {
        return endpointInjectionEnabled;
    }

    public void setEndpointInjectionEnabled(boolean endpointInjectionEnabled) {
        this.endpointInjectionEnabled = endpointInjectionEnabled;
    }

    public String getAuthIssuerUrl() {
        return authIssuerUrl;
    }

    public void setAuthIssuerUrl(String authIssuerUrl) {
        this.authIssuerUrl = authIssuerUrl;
    }

    public String getAuthOpenidConfigurationUrl() {
        return authOpenidConfigurationUrl;
    }

    public void setAuthOpenidConfigurationUrl(String authOpenidConfigurationUrl) {
        this.authOpenidConfigurationUrl = authOpenidConfigurationUrl;
    }

    public String getAuthOpenidIntrospectionUrl() {
        return authOpenidIntrospectionUrl;
    }

    public void setAuthOpenidIntrospectionUrl(String authOpenidIntrospectionUrl) {
        this.authOpenidIntrospectionUrl = authOpenidIntrospectionUrl;
    }

    public String getAuthOpenidTokenUrl() {
        return authOpenidTokenUrl;
    }

    public void setAuthOpenidTokenUrl(String authOpenidTokenUrl) {
        this.authOpenidTokenUrl = authOpenidTokenUrl;
    }

    public String getAuthOpenidRevokeUrl() {
        return authOpenidRevokeUrl;
    }

    public void setAuthOpenidRevokeUrl(String authOpenidRevokeUrl) {
        this.authOpenidRevokeUrl = authOpenidRevokeUrl;
    }

    public List<String> getExclusiveAuthScopes() {
        return exclusiveAuthScopes;
    }

    public void setExclusiveAuthScopes(List<String> exclusiveAuthScopes) {
        this.exclusiveAuthScopes = exclusiveAuthScopes;
    }

    public List<CorsConfigurationFilter> getCorsConfigurationFilters() {
        return corsConfigurationFilters;
    }

    public void setCorsConfigurationFilters(List<CorsConfigurationFilter> corsConfigurationFilters) {
        this.corsConfigurationFilters = corsConfigurationFilters;
    }

    public String getLoggingLevel() {
        return loggingLevel;
    }

    public void setLoggingLevel(String loggingLevel) {
        this.loggingLevel = loggingLevel;
    }

    public String getLoggingLayout() {
        return loggingLayout;
    }

    public void setLoggingLayout(String loggingLayout) {
        this.loggingLayout = loggingLayout;
    }

    public String getExternalLoggerConfiguration() {
        return externalLoggerConfiguration;
    }

    public void setExternalLoggerConfiguration(String externalLoggerConfiguration) {
        this.externalLoggerConfiguration = externalLoggerConfiguration;
    }

    public Boolean getDisableJdkLogger() {
        return disableJdkLogger;
    }

    public void setDisableJdkLogger(Boolean disableJdkLogger) {
        this.disableJdkLogger = disableJdkLogger;
    }

    public int getMaxCount() {
        return maxCount;
    }

    public void setMaxCount(int maxCount) {
        this.maxCount = maxCount;
        if(this.maxCount<=0) {
            this.maxCount = ApiConstants.DEFAULT_MAX_COUNT;
        }
    }

    public List<String> getAcrExclusionList() {
        return acrExclusionList;
    }

    public void setAcrExclusionList(List<String> acrExclusionList) {
        this.acrExclusionList = acrExclusionList;
    }

    public List<String> getUserExclusionAttributes() {
        return userExclusionAttributes;
    }

    public void setUserExclusionAttributes(List<String> userExclusionAttributes) {
        this.userExclusionAttributes = userExclusionAttributes;
    }

    public List<String> getUserMandatoryAttributes() {
        return userMandatoryAttributes;
    }

    public void setUserMandatoryAttributes(List<String> userMandatoryAttributes) {
        this.userMandatoryAttributes = userMandatoryAttributes;
    }

    public AgamaConfiguration getAgamaConfiguration() {
        return agamaConfiguration;
    }

    public void setAgamaConfiguration(AgamaConfiguration agamaConfiguration) {
        this.agamaConfiguration = agamaConfiguration;
    }

    public AuditLogConf getAuditLogConf() {
        return auditLogConf;
    }

    public void setAuditLogConf(AuditLogConf auditLogConf) {
        this.auditLogConf = auditLogConf;
    }

    public DataFormatConversionConf getDataFormatConversionConf() {
        return dataFormatConversionConf;
    }

    public void setDataFormatConversionConf(DataFormatConversionConf dataFormatConversionConf) {
        this.dataFormatConversionConf = dataFormatConversionConf;
    }

    public List<PluginConf> getPlugins() {
        return plugins;
    }

    public void setPlugins(List<PluginConf> plugins) {
        this.plugins = plugins;
    }

    public AssetMgtConfiguration getAssetMgtConfiguration() {
        return assetMgtConfiguration;
    }

    public void setAssetMgtConfiguration(AssetMgtConfiguration assetMgtConfiguration) {
        this.assetMgtConfiguration = assetMgtConfiguration;
    }

    @Override
    public String toString() {
        return "ApiAppConfiguration [serviceName=" + serviceName + ", configOauthEnabled=" + configOauthEnabled
                + ", disableLoggerTimer=" + disableLoggerTimer + ", disableAuditLogger=" + disableAuditLogger
                + ", customAttributeValidationEnabled=" + customAttributeValidationEnabled + ", acrValidationEnabled="
                + acrValidationEnabled + ", apiApprovedIssuer=" + apiApprovedIssuer + ", apiProtectionType="
                + apiProtectionType + ", apiClientId=" + apiClientId 
                + ", endpointInjectionEnabled=" + endpointInjectionEnabled + ", authIssuerUrl=" + authIssuerUrl
                + ", authOpenidConfigurationUrl=" + authOpenidConfigurationUrl + ", authOpenidIntrospectionUrl="
                + authOpenidIntrospectionUrl + ", authOpenidTokenUrl=" + authOpenidTokenUrl + ", authOpenidRevokeUrl="
                + authOpenidRevokeUrl + ", exclusiveAuthScopes=" + exclusiveAuthScopes + ", corsConfigurationFilters="
                + corsConfigurationFilters + ", loggingLevel=" + loggingLevel + ", loggingLayout=" + loggingLayout
                + ", externalLoggerConfiguration=" + externalLoggerConfiguration + ", disableJdkLogger="
                + disableJdkLogger + ", maxCount=" + maxCount + ", acrExclusionList=" + acrExclusionList
                + ", userExclusionAttributes=" + userExclusionAttributes + ", userMandatoryAttributes="
                + userMandatoryAttributes + ", agamaConfiguration=" + agamaConfiguration + ", auditLogConf="
                + auditLogConf + ", dataFormatConversionConf=" + dataFormatConversionConf + ", plugins=" + plugins
                + ", assetMgtConfiguration=" + assetMgtConfiguration + "]";
    }    
    
}

File: terraform-provider-jans/jans/api_app_configuration.go
package jans

import (
	"context"
	"fmt"
)

type AuditLogConf struct {
	Enabled          bool     `schema:"enabled" json:"enabled"`
	IgnoreHttpMethod []string `schema:"ignore_http_method" json:"ignoreHttpMethod"`
	HeaderAttributes []string `schema:"header_attributes" json:"headerAttributes"`
}

type DataFormatConversionConf struct {
	Enabled          bool     `schema:"enabled" json:"enabled"`
	IgnoreHttpMethod []string `schema:"ignore_http_method" json:"ignoreHttpMethod"`
}

type AgamaConfiguration struct {
	MandatoryAttributes []string `schema:"mandatory_attributes" json:"mandatoryAttributes"`
	OptionalAttributes  []string `schema:"optional_attributes" json:"optionalAttributes"`
}

type AssetDirMapping struct {
	Directory   string   `schema:"directory" json:"directory"`
	Type        []string `schema:"type" json:"type"`
	Description string   `schema:"description" json:"description"`
}

type AssetMgtConfiguration struct {
	AssetMgtEnabled                bool              `schema:"asset_mgt_enabled" json:"assetMgtEnabled"`
	AssetServerUploadEnabled       bool              `schema:"asset_server_upload_enabled" json:"assetServerUploadEnabled"`
	FileExtensionValidationEnabled bool              `schema:"file_extension_validation_enabled" json:"fileExtensionValidationEnabled"`
	ModuleNameValidationEnabled    bool              `schema:"module_name_validation_enabled" json:"moduleNameValidationEnabled"`
	AssetDirMappings               []AssetDirMapping `schema:"asset_dir_mappings" json:"assetDirMapping"`
}

// PersistenceConfiguration represents the persistence configuration
// of the Janssen server.
type ApiAppConfiguration struct {
	ConfigOauthEnabled               bool                      `schema:"config_oauth_enabled" json:"configOauthEnabled"`
	DisableLoggerTimer               bool                      `schema:"disable_logger_timer" json:"disableLoggerTimer"`
	DisableAuditLogger               bool                      `schema:"disable_audit_logger" json:"disableAuditLogger"`
	CustomAttributeValidationEnabled bool                      `schema:"custom_attribute_validation_enabled" json:"customAttributeValidationEnabled"`
	ArcValidationEnabled             bool                      `schema:"acr_validation_enabled" json:"acrValidationEnabled"`
	ApiApprovedIssuer                []string                  `schema:"api_approved_issuer" json:"apiApprovedIssuer"`
	ApiProtectionType                string                    `schema:"api_protection_type" json:"apiProtectionType"`
	ApiClientId                      string                    `schema:"api_client_id" json:"apiClientId"`
	ApiClientPassword                string                    `schema:"api_client_password" json:"apiClientPassword"`
	EndpointInjectionEnabled         bool                      `schema:"endpoint_injection_enabled" json:"endpointInjectionEnabled"`
	AuthIssuerUrl                    string                    `schema:"auth_issuer_url" json:"authIssuerUrl"`
	AuthOpenidConfigurationUrl       string                    `schema:"auth_openid_configuration_url" json:"authOpenidConfigurationUrl"`
	AuthOpenidIntrospectionUrl       string                    `schema:"auth_openid_introspection_url" json:"authOpenidIntrospectionUrl"`
	AuthOpenidTokenUrl               string                    `schema:"auth_openid_token_url" json:"authOpenidTokenUrl"`
	AuthOpenidRevokeUrl              string                    `schema:"auth_openid_revoke_url" json:"authOpenidRevokeUrl"`
	ExclusiveAuthScopes              []string                  `schema:"exclusive_auth_scopes" json:"exclusiveAuthScopes"`
	CorsConfigurationFilters         []CorsConfigurationFilter `schema:"cors_configuration_filters" json:"corsConfigurationFilters"`
	LoggingLevel                     string                    `schema:"logging_level" json:"loggingLevel"`
	LoggingLayout                    string                    `schema:"logging_layout" json:"loggingLayout"`
	ExternalLoggerConfiguration      string                    `schema:"external_logger_configuration" json:"externalLoggerConfiguration"`
	DisableJdkLogger                 bool                      `schema:"disable_jdk_logger" json:"disableJdkLogger"`
	MaxCount                         int                       `schema:"max_count" json:"maxCount"`
	UserExclusionAttributes          []string                  `schema:"user_exclusion_attributes" json:"userExclusionAttributes"`
	UserMandatoryAttributes          []string                  `schema:"user_mandatory_attributes" json:"userMandatoryAttributes"`
	AgamaConfiguration               AgamaConfiguration        `schema:"agama_configuration" json:"agamaConfiguration"`
	AuditLogConf                     AuditLogConf              `schema:"audit_log_conf" json:"auditLogConf"`
	DataFormatConversionConf         DataFormatConversionConf  `schema:"data_format_conversion_conf" json:"dataFormatConversionConf"`
	Plugins                          []PluginConf              `schema:"plugins" json:"plugins"`
	AssetMgtConfiguration            AssetMgtConfiguration     `schema:"asset_mgt_configuration" json:"assetMgtConfiguration"`
}

// GetApiAppConfiguration returns the current API configuration.
func (c *Client) GetApiAppConfiguration(ctx context.Context) (*ApiAppConfiguration, error) {

	token, err := c.getToken(ctx, "https://jans.io/oauth/config/properties.readonly")
	if err != nil {
		return nil, fmt.Errorf("failed to get token: %w", err)
	}

	ret := ApiAppConfiguration{}

	if err := c.get(ctx, "/jans-config-api/api/v1/api-config", token, &ret); err != nil {
		return nil, fmt.Errorf("get request failed: %w", err)
	}

	return &ret, nil
}

// PatchApiAppConfiguration  uses the provided list of patch requests to update
// the Janssen api configuration properties.
func (c *Client) PatchApiAppConfiguration(ctx context.Context, patches []PatchRequest) (*ApiAppConfiguration, error) {

	if len(patches) == 0 {
		return c.GetApiAppConfiguration(ctx)
	}

	orig, err := c.GetApiAppConfiguration(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get app configuration: %w", err)
	}

	updates, err := createPatchesDiff(orig, patches)
	if err != nil {
		return nil, fmt.Errorf("failed to create patches: %w", err)
	}

	if len(updates) == 0 {
		return c.GetApiAppConfiguration(ctx)
	}

	token, err := c.getToken(ctx, "https://jans.io/oauth/config/properties.write")
	if err != nil {
		return nil, fmt.Errorf("failed to get token: %w", err)
	}

	if err := c.patch(ctx, "/jans-config-api/api/v1/api-config", token, updates); err != nil {
		return nil, fmt.Errorf("patch request failed: %w", err)
	}

	return c.GetApiAppConfiguration(ctx)
}
